Testschains: Many regtests with different genesis and default datadir

jtimon opened the pull request:
And bip70 name.
And on demand (meaning you don't have to mine a genesis block for it to pass minimal pow).

This allows the daemon to create a new regtest-like chains on demand.
The hash of the genesis block is dependent on the "petname" of the chain.
Examples: "regtest2, custom, chain_2, aaa, bbbb"

The hash of the genesis block could depend on more things.
For example, for signet chains, it could depend on the signet_blockscript in the case of BIP325 (signet) https://github.com/bitcoin/bitcoin/pull/16411/

In fact, #16411 could be simplified if this was merged first, for the genesis block wouldn't need to be mined anymore (which requires a special case in grindblock which in turn requires CreateSignetGenesisBlock to be exposed in chainparams). So I guess perhaps signet could be counted as a use case, or perhaps only part of this.

But why would somebody want more than one regtest?

I'm personally using it for doing cross-chain payments in lightning, see https://github.com/jtimon/multi-ln-demo/blob/master/conf/bitcoind.conf
I would like to work on what I call "cross chain trampoline payments" and I plan to keep using it for that too.

I imagine other developers could find this useful for other developments involving bitcoin-like chains.
For example atomic swaps or submarine swaps.

Of course, an alternative for these use cases is to use other regtests from other bitcoin-like networks, for example litecoin regtest or liquid regtest.

Another use case is creating temporal testnets for testing upcoming features.
For example, had this been in place before segwit, when "segwitnet" (was that its name) was created, it could have simply been some shared configuration instead of an additional hardoced chainparams.
Something like:

```
[segwitnet]
segwitheight=0
rpcport=18555
port=18556
...
```

We're not going to use it for a segwitnet now, obviously, but perhaps for a taprootnet or something.
Perhaps #17032 would be needed too for this use case in particular to be more useful though, or at least make that for some of the fields that are different between testnet3 and regtest.

jtimon committed:
Testchains: Qt: Introduce custom chains with different:

1) genesis block hash
2) default datadir
3) chain name (bip70)

all directly or indirectly configured with the -chain option

...whose constructor reads params from regular arguments (like regtests)...

Qt: Add a default purple and title for unkown chains

instagibbs committed:
9102: Really don't validate genesis block

jtimon committed:
Config: Allow any chain/section name

jtimon committed:
Test: Custom chain genesis' are deterministic from the name

jtimon commented:
Mhmm, I'm not sure why the new test works. In ```initialize_datadir()``` in test/test_framework/util.py it says:

```
f.write("{}=1\n".format(chain_name_conf_arg))
```

which would result in 
```
chain_5=1
```
for example instead of resulting in 

```
chain=chain_5
```

as it should and would with:
```
        f.write("chain={}\n".format(chain_name_conf_arg))
```

If I do the change, it still works, but then feature_config_args fails in the test that is run with mainnet (something that #17032 fixes, btw).

Any ideas on why is it working?

jtimon committed:
Chainparams: Test: Make is_test_chain configurable for regtest

jtimon committed:
Test: select chain using -chain=regtest instead of regtest=1

jtimon commented:
I still don't understand why the previous version without making is_test_chain customizable worked in my computer and in all travis builds except one. But now it should work for all the builds.
If anyone is able to do explain why or do a version without that commit, it is more than welcomed.

jtimon commented:
Rebased

JeremyRubin commented:
Can you write a more detailed PR description motivating what use cases require this?

jtimon commented:
Sure, done. Thanks for the suggestion, it is true that the motivation wasn't well explained at all.
I think it is much better now, but please suggest more changes to the description if you think of anything else to improve.

JeremyRubin commented:
Can you explain more why it's worthwile to review this and not just signet?

I understand it makes it a bit easier to make a genesis block for signet, but how hard is that?

jtimon commented:
This would simplify signet.
Right now with signet, you need to call grindblock for the genesis block.
It also makes the code more complicated.
If signet gets merged first, I can rebase this as a simplification of signet.
Ideally both would be reviewed.
Another option would be to simply modify #16411 , but I don't think @kallewoof will be open to that unless he knows for sure that skipping pow validation in the genesis block is fine with people.
Although I guess I should let him speak for himself.

JeremyRubin commented:
I'd be interested to hear @kallewoof's thoughts on if this PR is good for signet or not -- it seems like that having the signet stuff merged would make this "nice to have", but not enabling something new (as you can use signets instead of regtests).

jtimon commented:
Yeah, only one configurable class is necessary. Even making not signet
regtests should be as simple as making is_signet configurable.
Whether that class is called CSignetChainparams or CCustomChainparams
shouldn't be very important IMO.
And if it was the other way around, signet could just be adding those
parameters to the existing class.
I can produce 2 branches with this and signet. One with this first and the
other with signet first.
Although if I do both of them the end result will be the same in both. But
that may be useful to see what I have in mind. And to discuss things that
could be different.
If there's interest, I'm happy to do them.




On Fri, Dec 20, 2019, 19:27 Jeremy Rubin <notifications@github.com> wrote:

> I'd be interested to hear @kallewoof <https://github.com/kallewoof>'s
> thoughts on if this PR is good for signet or not -- it seems like that
> having the signet stuff merged would make this "nice to have", but not
> enabling something new (as you can use signets instead of regtests).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/17037?email_source=notifications&email_token=AAHWGSVIANX3MEUBC2IWNTTQZUFC3A5CNFSM4I453GF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHNYXCY#issuecomment-568036235>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHWGSSLVDAHLLSJXQLPG4TQZUFC3ANCNFSM4I453GFQ>
> .
>

kallewoof commented:
I don't have strong opinions on this PR in particular, as @JeremyRubin stated, we can use multiple signets to get approximately the same effect as multiple custom-named regtests. I don't think this would be detrimental to signet, though.

As a sidenote, I would love to remove the grind genesis proof of work part from signet, but unfortunately this was discussed and people didn't like it so it was removed. I don't know if people will have a different opinion for this PR, but maybe with more feedback, we get a more complete picture on the topic.

Anyway, I'm pretty much neutral.

jtimon commented:
Rebased.

Regarding -signet_genesisnonce , as far as I understand the hesitation to remove it is having to touch consensus code. But the changes are minimal, it is just the first commit in this PR.
As a reminder, the main reason for not merging that commit alone in its own PR, was that it wasn't tested (this PR does test it, as  #16411 would if it included that same commit and removed -signet_genesisnonce), see https://github.com/bitcoin/bitcoin/pull/9102#issuecomment-259097238

jtimon commented:
Rebased

jtimon commented:
Closing due to lack of interest.
